library/bignum.c

```diff
@@ -2205,23 +2212,23 @@ int mbedtls_mpi_gen_prime( mbedtls_mpi *X, size_t nbits, int dh_flag,
 
     n = BITS_TO_LIMBS( nbits );
 
+new_number:
     MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( X, n * ciL, f_rng, p_rng ) );
+    /* make sure generated number is at least (n-1)+0.5 bits */
+    if( X->p[n-1] < CEIL_MAXUINT_DIV_SQRT2 ) goto new_number;
 
-    k = mbedtls_mpi_bitlen( X );
-    if( k > nbits ) MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( X, k - nbits + 1 ) );
-
-    mbedtls_mpi_set_bit( X, nbits-1, 1 );
-
+    k = n * biL;
+    if( k > nbits ) MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( X, k - nbits ) );
```

Comment:
There is a useless yet possible corner case in which this operation breaks the property that `X >= 2^(nbits - 0.5)`, namely if `n < biL`. To verify, you can temporarily compute `X^2` and assert `mbedtls_mpi_bitlen( &X ) >= 2 * nbits`, which is not always true (you can e.g. run a few iterations of `./programs/pkey/dh_gen_prime bits=3` to see this).